# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: layout/generic/nsFrame.cpp
# Commit: b3f0e5c7874d
# Full Hash: b3f0e5c7874dbd470d96e476e116740e4ca35925
# Author: David Zbarsky <dzbarsky@gmail.com>
# Date: 2012-08-06 14:40:43
# Regressor Bug: 779684
# File Overlap Count: 1
# Description:
#   Bug 779684: Create a stack-based nsFrameIterator class out of nsIFrameTraversal and nsIFrameEnumerator r=roc
# ==============================================================================

diff -r 4bc7154190dd -r b3f0e5c7874d layout/generic/nsFrame.cpp
--- a/layout/generic/nsFrame.cpp	Sun Aug 05 22:55:18 2012 -0400
+++ b/layout/generic/nsFrame.cpp	Sun Aug 05 23:00:55 2012 -0400
@@ -52,7 +52,7 @@
 #include "nsCSSPseudoElements.h"
 #include "nsCSSFrameConstructor.h"
 
-#include "nsFrameTraversal.h"
+#include "nsFrameIterator.h"
 #include "nsStyleChangeList.h"
 #include "nsIDOMRange.h"
 #include "nsRange.h"
@@ -5782,16 +5782,12 @@
       //resultFrame is not a block frame
       result = NS_ERROR_FAILURE;
 
-      nsCOMPtr<nsIFrameEnumerator> frameTraversal;
-      result = NS_NewFrameTraversal(getter_AddRefs(frameTraversal),
-                                    aPresContext, resultFrame,
-                                    ePostOrder,
-                                    false, // aVisual
-                                    aPos->mScrollViewStop,
-                                    false     // aFollowOOFs
-                                    );
-      if (NS_FAILED(result))
-        return result;
+      PRUint32 flags = FrameIteratorFlags::FLAG_NONE;
+      if (aPos->mScrollViewStop) {
+        flags |= FrameIteratorFlags::FLAG_LOCK_SCROLL;
+      }
+      nsFrameIterator frameTraversal(aPresContext, resultFrame,
+                                     ePostOrder, flags);
       nsIFrame *storeOldResultFrame = resultFrame;
       while ( !found ){
         nsPoint point;
@@ -5868,21 +5864,21 @@
         if (aPos->mDirection == eDirNext && (resultFrame == nearStoppingFrame))
           break;
         //always try previous on THAT line if that fails go the other way
-        frameTraversal->Prev();
-        resultFrame = frameTraversal->CurrentItem();
+        frameTraversal.Prev();
+        resultFrame = frameTraversal.CurrentItem();
         if (!resultFrame)
           return NS_ERROR_FAILURE;
       }
 
       if (!found){
         resultFrame = storeOldResultFrame;
-        result = NS_NewFrameTraversal(getter_AddRefs(frameTraversal),
-                                      aPresContext, resultFrame,
-                                      eLeaf,
-                                      false, // aVisual
-                                      aPos->mScrollViewStop,
-                                      false     // aFollowOOFs
-                                      );
+
+        PRUint32 flags = FrameIteratorFlags::FLAG_NONE;
+        if (aPos->mScrollViewStop) {
+          flags |= FrameIteratorFlags::FLAG_LOCK_SCROLL;
+        }
+        frameTraversal = nsFrameIterator(aPresContext, resultFrame,
+                                         eLeaf, flags);
       }
       while ( !found ){
         nsPoint point(aPos->mDesiredX, 0);
@@ -5913,8 +5909,8 @@
         if (aPos->mDirection == eDirNext && (resultFrame == farStoppingFrame))
           break;
         //previous didnt work now we try "next"
-        frameTraversal->Next();
-        nsIFrame *tempFrame = frameTraversal->CurrentItem();
+        frameTraversal.Next();
+        nsIFrame *tempFrame = frameTraversal.CurrentItem();
         if (!tempFrame)
           break;
         resultFrame = tempFrame;
@@ -6638,23 +6634,22 @@
         return NS_ERROR_FAILURE; //we are done. cannot jump lines
     }
 
-    nsCOMPtr<nsIFrameEnumerator> frameTraversal;
-    result = NS_NewFrameTraversal(getter_AddRefs(frameTraversal),
-                                  presContext, traversedFrame,
-                                  eLeaf,
-                                  aVisual && presContext->BidiEnabled(),
-                                  aScrollViewStop,
-                                  true     // aFollowOOFs
-                                  );
-    if (NS_FAILED(result))
-      return result;
+    PRUint32 flags = FrameIteratorFlags::FLAG_FOLLOW_OUT_OF_FLOW;
+    if (aScrollViewStop) {
+      flags |= FrameIteratorFlags::FLAG_LOCK_SCROLL;
+    }
+    if (aVisual && presContext->BidiEnabled()) {
+      flags |= FrameIteratorFlags::FLAG_VISUAL;
+    }
+    nsFrameIterator frameTraversal(presContext, traversedFrame,
+                                   eLeaf, flags);
 
     if (aDirection == eDirNext)
-      frameTraversal->Next();
+      frameTraversal.Next();
     else
-      frameTraversal->Prev();
-
-    traversedFrame = frameTraversal->CurrentItem();
+      frameTraversal.Prev();
+
+    traversedFrame = frameTraversal.CurrentItem();
     if (!traversedFrame)
       return NS_ERROR_FAILURE;
     traversedFrame->IsSelectable(&selectable, nullptr);